### DISCUSSION TITLE
Remove extra white space in className at build time

### DISCUSSION THREAD

**Discussion by Nases** [vercel/next.js#23305] [Ideas]
### Describe the feature you'd like to request

Remove extra white spaces in `className`'s at build time. 

I usually define `className`'s like below for developer friendliness;
```javascript
<button
      className={`
        inline-flex justify-center font-medium rounded-md
        ${sizes[size] || defaultSize}
        ${colors[color] || defaultColor}
        ${disabled ? 'opacity-50 cursor-default' : ''}
        ${className || ''}
      `}
    >
{children}
</button>
```
In production HTML looks like this
```html
<button type="button" class="
        inline-flex justify-center font-medium rounded-md
        py-2 px-4 text-sm
        shadow text-primary-600 bg-white hover:bg-gray-50 hover:text-primary-700 active:text-primary-900
        
        mt-3 w-full
      ">Cancel</button>
```

### Describe the solution you'd like

Add Next.js's default webpack config extra process in order to remove extra white spaces in `className`'s at build time.

### Describe alternatives you've considered

I've tried removing in the client side although it consumes unnecessary memory from our customers.

**Comment by nikoladev** [comment]
You might have already looked into this, but [classnames][cn] does this. If you're looking for an even lighter version I have made [a simpler version of classnames][dcn] that only accepts arrays. [This is literally all the code for it][code].

[cn]: https://github.com/JedWatson/classnames/
[dcn]: https://github.com/nikoladev/libraries/tree/main/packages/dumb-classnames
[code]: https://github.com/nikoladev/libraries/blob/main/packages/dumb-classnames/src/index.ts

**Reply by Nases** [reply]
I've looked into [classnames](https://www.npmjs.com/package/classnames) & [clsx](https://www.npmjs.com/package/clsx) although it would run every time that component renders.

Wouldn't it better to just remove extra spaces in build time? Not sure if there are edge cases when the extra spaces are actually useful.

**Reply by nikoladev** [reply]
I'm not sure if what I'm about to say is true, but you could also create the class names outside of the components and then they will only be calculated once on the client and not on every re-render.

Something like this:

```js
const longClassName = classNames([
  'inline-flex justify-center font-medium rounded-md',
  sizes[size] || defaultSize,
  colors[color] || defaultColor,
  disabled ? 'opacity-50 cursor-default' : '',
  className || '',
])

function Component () {
  return (
    <button
      className={longClassName}
    >
      Cancel
    </button>
}
```

**Reply by Nases** [reply]
I grab `size`, `color`, `className` etc from component's `props`, sadly I can't use it like that.

**Reply by nikoladev** [reply]
Would it be possible to use [`getStaticProps`][gsp] for that? Anything within that function only gets run at build time, so the class names (and any other props) can be determined at build time and passed to the page as unchanging props.

```js

export default function Page ({ buttonClassName }) {
  return (
    <button
      className={buttonClassName}
    >
      Cancel
    </button>
}

export async function getStaticProps(context) {
  // ⚠️ add the logic here for determining what the class names need to be

  const buttonClassName = classNames([
    'inline-flex justify-center font-medium rounded-md',
    sizes[size] || defaultSize,
    colors[color] || defaultColor,
    disabled ? 'opacity-50 cursor-default' : '',
    className || '',
  ])

  return {
    props: { buttonClassName }, // will be passed to the page component as props
  }
}

```

[gsp]: https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation
